src/meta/src/stream/scale.rs

```diff
@@ -1474,3 +1489,255 @@ where
+                        .difference(&fragment_parallel_unit_ids)
+                        .sorted()
+                        .cloned()
+                        .collect_vec();
+
+                    let to_shrink_parallel_units = fragment_parallel_unit_ids
+                        .difference(&target_parallel_unit_ids)
+                        .sorted()
+                        .cloned()
+                        .collect_vec();
+
+                    target_plan.insert(
+                        fragment_id,
+                        ParallelUnitReschedule {
+                            added_parallel_units: to_expand_parallel_units,
+                            removed_parallel_units: to_shrink_parallel_units,
+                        },
+                    );
+                }
+            }
+        }
+
+        Self::clone_no_shuffle_plan(&mut target_plan, &actor_map, &fragment_map)?;
+
+        target_plan.drain_filter(|_, plan| {
+            plan.added_parallel_units.is_empty() && plan.removed_parallel_units.is_empty()
+        });
+
+        Ok(target_plan)
+    }
+
+    fn clone_no_shuffle_plan(
+        target_plan: &mut HashMap<FragmentId, ParallelUnitReschedule>,
+        actor_map: &HashMap<ActorId, StreamActor>,
+        fragment_map: &HashMap<FragmentId, Fragment>,
+    ) -> MetaResult<()> {
+        let mut no_shuffle_source_fragment_ids = HashSet::new();
+        let mut no_shuffle_target_fragment_ids = HashSet::new();
+
+        let mut fragment_dispatcher_map = HashMap::new();
+
+        Self::build_no_shuffle_index(
+            actor_map,
+            &mut no_shuffle_source_fragment_ids,
+            &mut no_shuffle_target_fragment_ids,
+            &mut fragment_dispatcher_map,
+        );
+
+        let mut queue: VecDeque<FragmentId> = target_plan.keys().cloned().collect();
+
+        // We trace the upstreams of each downstream under the hierarchy until we reach the top for
+        // every no_shuffle relation.
+        while let Some(fragment_id) = queue.pop_front() {
+            if !no_shuffle_target_fragment_ids.contains(&fragment_id)
+                && !no_shuffle_source_fragment_ids.contains(&fragment_id)
+            {
+                continue;
+            }
+
+            // for upstream
+            for upstream_fragment_id in &fragment_map
+                .get(&fragment_id)
+                .unwrap()
+                .upstream_fragment_ids
+            {
+                if !no_shuffle_source_fragment_ids.contains(upstream_fragment_id) {
+                    continue;
+                }
+
+                let plan = target_plan.get(&fragment_id).unwrap();
+
+                if let Some(upstream_plan) = target_plan.get(upstream_fragment_id) {
+                    if upstream_plan != plan {
+                        bail!("Inconsistent NO_SHUFFLE plan, check target worker ids of fragment {} and {}", fragment_id, upstream_fragment_id);
```

Comment:
Is this an internal error, or can occur if users specify multiple fragments with NoShuffle exchange in their inputs considering the instablity of `HashMap::iter`.

Reply:
The parallel unit ids in the plan that have been added to the `target_plan` are all sorted. If a user provides the same `include_workers` and `exclude_workers` for the fragment id in the NO_SHUFFLE relationship, the resulting plan will certainly be the same. Therefore, we can check the accuracy of the workers parameter provided by the user in this way.

Reply:
Do we ensure this for singleton fragment as well?

Reply:
You make a good point! I'll go make the changes right away!

Reply:
It'll be nice if we can cover this with a simulation test. 🫡

Reply:
I've added simulated tests for both expected successes and failures. 😃